snap/snapcraft.yaml

```diff
@@ -21,130 +21,128 @@ plugs:
     bus: system
     name: io.openthread.BorderRouter.wpan0
 
+  # Uses system-files snap interface to access the gpiochip0
+  dev-gpiochip0:
+    interface: system-files
+    read:
+      - /sys/devices/platform/soc/fe200000.gpio/gpiochip0/dev
```

Comment:
`fe200000` is this expected to be the same on all Pis and distros?

Reply:
This seems to be related to the memory address of the `/dev/gpiochip0` in the RPi4/BCM2711 case scenario. I'm going to confirm it.

Reply:
On a RPi3, running Raspbian 10:
```
$ gpiodetect
gpiochip0 [pinctrl-bcm2835] (54 lines)
gpiochip1 [brcmvirt-gpio] (2 lines)
gpiochip2 [raspberrypi-exp-gpio] (8 lines)
```

And the sys path is `/sys/devices/platformorm/soc/3f200000.gpio/gpiochip0/dev`

Reply:
It's definely Hardware dependent, for the raspberry RPi4 runnin Raspbian 12 is the same as for Ubuntu 22.04 

```bash
pi4@alsvid:~ $ dmesg | grep gpio 
[    8.575625] rpi-gpiomem fe200000.gpiomem: window base 0xfe200000 size 0x00001000
[    8.654533] rpi-gpiomem fe200000.gpiomem: initialised 1 regions as /dev/gpiomem
```
`fe200000` is the upper limit of the range for the memory address for the gpio controller on `/dev/gpiochip0`.

```bash
pi4@alsvid:~ $ sudo ls /sys/firmware/devicetree/base/soc/
'#address-cells'        dma-ranges	       hdmi@7ef05700   i2c@7ef09500		       name		     ranges	       sound	        usb@7e980000
 aux@7e215000	        dpi@7e208000	       hvs@7e400000    i2s@7e203000		       phandle		     rng@7e104000      spi@7e204000     vec@7ec13000
 avs-monitor@7d5d2000   dsi@7e209000	       i2c0mux	       interrupt-controller@40041000   pixelvalve@7e206000   serial@7e201000   spi@7e204600     watchdog@7e100000
 axiperf	        dsi@7e700000	       i2c@7e205000    interrupt-controller@7ef00100   pixelvalve@7e207000   serial@7e201400   spi@7e204800
 clock@7ef00000         fb		       i2c@7e205600    local_intc@40000000	       pixelvalve@7e20a000   serial@7e201600   spi@7e204a00
 compatible	        firmware	       i2c@7e205800    mailbox@7e00b840		       pixelvalve@7e216000   serial@7e201800   spi@7e204c00
 cprman@7e101000        firmwarekms@7e600000   i2c@7e205a00    mailbox@7e00b880		       pixelvalve@7ec12000   serial@7e201a00   spi@7e215080
 csi@7e800000	        gpio@7e200000	       i2c@7e205c00    mmc@7e202000		       power		     serial@7e215040   spi@7e2150c0
 csi@7e801000	        gpiomem		       i2c@7e804000    mmc@7e300000		       pwm@7e20c000	    '#size-cells'      timer@7e003000
 dma@7e007000	        hdmi@7ef00700	       i2c@7ef04500    mmcnr@7e300000		       pwm@7e20c800	     smi@7e600000      txp@7e004000
pi4@alsvid:~ $
```

And the `7e200000` is the bottom one.

Reply:
Ok, so this isn't resolved. The snap should work on all arm64 Pis. That means we need to find out the addresses for various BCM chips and make it possible to control GPIO on all.

Reply:
The problem with grant access to the parent directory is that it's too permissive/wide.

```bash
$ ls -l /sys/devices/platform/soc/
total 0
lrwxrwxrwx 1 root root    0 dez 31  1969 driver -> ../../../bus/platform/drivers/simple-pm-bus
-rw-r--r-- 1 root root 4096 jan 25 08:30 driver_override
drwxr-xr-x 4 root root    0 dez 31  1969 fd5d2000.avs-monitor
drwxr-xr-x 3 root root    0 dez 31  1969 fe004000.txp
drwxr-xr-x 4 root root    0 dez 31  1969 fe007000.dma
drwxr-xr-x 9 root root    0 dez 31  1969 fe00b840.mailbox
drwxr-xr-x 3 root root    0 dez 31  1969 fe00b880.mailbox
drwxr-xr-x 5 root root    0 dez 31  1969 fe100000.watchdog
drwxr-xr-x 3 root root    0 dez 31  1969 fe101000.cprman
drwxr-xr-x 3 root root    0 dez 31  1969 fe104000.rng
drwxr-xr-x 5 root root    0 dez 31  1969 fe200000.gpio
drwxr-xr-x 3 root root    0 dez 31  1969 fe200000.gpiomem
drwxr-xr-x 4 root root    0 dez 31  1969 fe201000.serial
drwxr-xr-x 4 root root    0 dez 31  1969 fe204000.spi
drwxr-xr-x 3 root root    0 dez 31  1969 fe206000.pixelvalve
drwxr-xr-x 3 root root    0 dez 31  1969 fe207000.pixelvalve
drwxr-xr-x 3 root root    0 dez 31  1969 fe20a000.pixelvalve
drwxr-xr-x 3 root root    0 dez 31  1969 fe215000.aux
drwxr-xr-x 3 root root    0 dez 31  1969 fe216000.pixelvalve
drwxr-xr-x 4 root root    0 dez 31  1969 fe300000.mmcnr
drwxr-xr-x 3 root root    0 dez 31  1969 fe400000.hvs
drwxr-xr-x 4 root root    0 dez 31  1969 fe804000.i2c
drwxr-xr-x 6 root root    0 dez 31  1969 fe980000.usb
drwxr-xr-x 3 root root    0 dez 31  1969 fec12000.pixelvalve
drwxr-xr-x 3 root root    0 dez 31  1969 fef00000.clock
drwxr-xr-x 8 root root    0 dez 31  1969 fef00700.hdmi
drwxr-xr-x 4 root root    0 dez 31  1969 fef04500.i2c
drwxr-xr-x 8 root root    0 dez 31  1969 fef05700.hdmi
drwxr-xr-x 4 root root    0 dez 31  1969 fef09500.i2c
drwxr-xr-x 3 root root    0 dez 31  1969 ff800000.local_intc
-r--r--r-- 1 root root 4096 jan 25 08:30 modalias
lrwxrwxrwx 1 root root    0 jan 25 08:30 of_node -> ../../../firmware/devicetree/base/soc
drwxr-xr-x 2 root root    0 jan 25 08:30 power
drwxr-xr-x 8 root root    0 dez 31  1969 soc:firmware
drwxr-xr-x 3 root root    0 dez 31  1969 soc:power
lrwxrwxrwx 1 root root    0 dez 31  1969 subsystem -> ../../../bus/platform
-rw-r--r-- 1 root root 4096 dez 31  1969 uevent
```

Reply:
| Model | BCM Chip | gpiochip0 address | Notes |
|---|---|---|---|
| Raspberry Pi 3 Model B Rev 1.2 | bcm2835 | 3f200000 | armv7 but same chip has been used on some arm64 3B+ |
| Raspberry Pi 4 Model B Rev 1.1 | bcm2711 | fe200000 |  |
| Raspberry Pi 4 Model B Rev 1.5 | bcm2711 | fe200000 |  |
|  |  |  |  |

Reply:
[This commit](https://github.com/canonical/matter-pi-gpio-commander/pull/42/commits/039b537b47a0fc873fabf789fd050d2a1cb077f5) solves this problem.

We just need to double-check on RPi5 and see if follows the same directory structure but with the`/dev/gpiochip4` instead.